Sources/SourceKitLSP/TestDiscovery.swift

```diff
@@ -462,6 +475,105 @@ extension TestItem {
+  /// If we think of the enclosing class or struct as the test suite, then extensions on that class or struct should be
+  /// additions to that suite, just like extensions on types are, from the user's perspective, transparently added to
+  /// their type.
+  ///
+  /// This method walks the `TestItem` tree produced by the test scanners and merges in the tests defined in extensions
+  /// into the `TestItem` that represents the type definition.
+  ///
+  /// This causes extensions to be merged into their type's definition if the type's definition exists in the list of
+  /// test items. If the type's definition is not a test item in this collection, the first extension of that type will
+  /// be used as the primary test location.
+  ///
+  /// For example if there are two files
+  ///
+  /// FileA.swift
+  /// ```swift
+  /// @Suite struct MyTests {
+  ///   @Test func oneIsTwo {}
+  /// }
+  /// ```
+  ///
+  /// FileB.swift
+  /// ```swift
+  /// extension MyTests {
+  ///   @Test func twoIsThree() {}
+  /// }
+  /// ```
+  ///
+  /// Then `workspace/tests` will return
+  /// - `MyTests` (FileA.swift:1)
+  ///   - `oneIsTwo`
+  ///   - `twoIsThree`
+  ///
+  /// And `textDocument/tests` for FileB.swift will return
+  /// - `MyTests` (FileB.swift:1)
+  ///   - `twoIsThree`
+  ///
+  /// A node's parent is identified by the node's ID with the last component dropped.
+  func mergingTestsInExtensions() -> [TestItem] {
+    var itemDict: [String: TestItem] = [:]
+    for item in self {
+      if var rootItem = itemDict[item.id] {
+        // If we've encountered an extension first, and this is the
+        // type declaration, then use the type declaration TestItem
+        // as the root item.
+        if rootItem.isExtension && !item.isExtension {
+          var newItem = item
+          newItem.children += rootItem.children
+          rootItem = newItem
+        } else {
+          rootItem.children += item.children
+        }
+
+        itemDict[item.id] = rootItem
+      } else {
+        itemDict[item.id] = item
+      }
+    }
+
+    if itemDict.isEmpty {
+      return []
+    }
+
+    for item in self {
+      let parentID = item.id.components(separatedBy: "/").dropLast().joined(separator: "/")
+      // If the parent exists, add the current item to its children and remove it from the root
+      if var parent = itemDict[parentID] {
+        parent.children.append(item)
+        itemDict[parent.id] = parent
+        itemDict[item.id] = nil
+      }
+    }
+
+    // Filter out the items that have been merged into their parents, sorting the tests by location.
+    // TestItems not in extensions should be priotitized first.
+    var sortedItems = itemDict.values.compactMap { $0 }.sorted {
+      ($0.location.uri != $1.location.uri && $0.isExtension != $1.isExtension) ? !$0.isExtension : ($0.location < $1.location)
+    }
+
+    sortedItems = sortedItems.map {
+      guard !$0.children.isEmpty else {
+        return $0
+      }
```

Comment:
Would it make sense to keep track of the IDs for which we merged items? Then we wouldn’t need to call `$0.children.mergingTestsInExtensions()` for the test items that weren’t touched, which would be more performant since I expect in most cases there is no extension for a test class or suite. Or am I missing something?